ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/storage/keyvalue/RocksDbStorageProvider.java

```diff
@@ -46,34 +56,48 @@ public static StorageProvider create(
       final MetricsSystem metricsSystem,
       final long worldStatePreimageCacheSize)
       throws IOException {
-    if (rocksDbConfiguration.useColumns()) {
-      return createSegmentedProvider(
-          rocksDbConfiguration, metricsSystem, worldStatePreimageCacheSize);
+
+    final Path databaseDir = rocksDbConfiguration.getDatabaseDir();
+    final boolean databaseExists = databaseDir.resolve("IDENTITY").toFile().exists();
+    final int databaseVersion;
+    if (databaseExists) {
+      databaseVersion = DatabaseMetadata.fromDirectory(databaseDir).getVersion();
+      LOG.info("Existing database detected. Version {}", databaseVersion);
     } else {
-      return createUnsegmentedProvider(
-          rocksDbConfiguration, metricsSystem, worldStatePreimageCacheSize);
+      databaseVersion = CURRENT_VERSION;
+      LOG.info("No existing database detected. Using version {}", databaseVersion);
+      Files.createDirectories(databaseDir);
+      new DatabaseMetadata(databaseVersion).writeToDirectory(databaseDir);
+    }
+
+    try {
+      return PROVIDERS_BY_VERSION
```

Comment:
If we open a DB of the wrong version what happens?  Does it fail sanity checks or does it fail catastrophically and destroy data?

Reply:
What does rocks db do when you try to load a database with options/schema it wasn't created with?

Reply:
I'm not sure, we should try and see if any data changes.

Reply:
Depends on the changes.  If you try to open a database for writing without specifying all the columns used in that database it will throw an exception and refuse to open.  So opening a columnar db with our legacy non-columnar format will just refuse to open, but I doubt that's true of all potentially incompatible changes we'd make.

Reply:
Response to my question on RocksDB dev group:

> I cannot think of an option that could not be changed back. Your right that some options have persistent impact on DB but it should not prevent the DB to be successfully loaded after the option change. In many cases, if the option has a persistent impact it is remembered in the properties of the SST file, so the SST file could still be read properly even if the option is not set anymore. The newly generated files, however, will follow the newest values of options. For "optimize filters for hits" option in particular, you might need to do a manual compaction on all db if you want to see the optimization impact of the option change immediately.

We're not at risk of accidentally breaking the database by loading it with the wrong options. If we do make a developer error and the database is being loaded with the wrong version, it will either fail like @ajsutton mentioned for columns, which is a deep structural property, or load with performance characteristics we're not expecting. At least this case is something that doesn't destroy the existing data while we push out a fix.